<h1>Содержание<span class="tocSkip"></span></h1>
<div class="toc"><ul class="toc-item"><li><span><a href="#Импорт-необходимых-библиотек" data-toc-modified-id="Импорт-необходимых-библиотек-1"><span class="toc-item-num">1&nbsp;&nbsp;</span>Импорт необходимых библиотек</a></span></li><li><span><a href="#Загрузка-данных" data-toc-modified-id="Загрузка-данных-2"><span class="toc-item-num">2&nbsp;&nbsp;</span>Загрузка данных</a></span></li><li><span><a href="#Умножение-матриц" data-toc-modified-id="Умножение-матриц-3"><span class="toc-item-num">3&nbsp;&nbsp;</span>Умножение матриц</a></span></li><li><span><a href="#Алгоритм-преобразования" data-toc-modified-id="Алгоритм-преобразования-4"><span class="toc-item-num">4&nbsp;&nbsp;</span>Алгоритм преобразования</a></span></li><li><span><a href="#Проверка-алгоритма" data-toc-modified-id="Проверка-алгоритма-5"><span class="toc-item-num">5&nbsp;&nbsp;</span>Проверка алгоритма</a></span></li><li><span><a href="#Вывод" data-toc-modified-id="Вывод-6"><span class="toc-item-num">6&nbsp;&nbsp;</span>Вывод</a></span></li><li><span><a href="#Чек-лист-проверки" data-toc-modified-id="Чек-лист-проверки-7"><span class="toc-item-num">7&nbsp;&nbsp;</span>Чек-лист проверки</a></span></li></ul></div>

# Защита персональных данных клиентов

Вам нужно защитить данные клиентов страховой компании «Хоть потоп». Разработайте такой метод преобразования данных, чтобы по ним было сложно восстановить персональную информацию. Обоснуйте корректность его работы.

Нужно защитить данные, чтобы при преобразовании качество моделей машинного обучения не ухудшилось. Подбирать наилучшую модель не требуется.

## Импорт необходимых библиотек

In [1]:
import pandas as pd
import numpy as np
from sklearn.metrics import mean_squared_error
from sklearn.model_selection import train_test_split
from sklearn.linear_model import LinearRegression
from sklearn.metrics import accuracy_score
import warnings
warnings.filterwarnings('ignore')
from sklearn.metrics import r2_score
from tqdm.notebook import tqdm
import matplotlib.pyplot as plt
import seaborn as sns

## Загрузка данных

In [2]:
df = pd.read_csv('/datasets/insurance.csv')
df.head()

Пол  Возраст  Зарплата  Члены семьи  Страховые выплаты
0    1     41.0   49600.0            1                  0
1    0     46.0   38000.0            1                  1
2    0     29.0   21000.0            0                  0
3    0     21.0   41700.0            2                  0
4    1     28.0   26100.0            0                  0

In [3]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 5000 entries, 0 to 4999
Data columns (total 5 columns):
 #   Column             Non-Null Count  Dtype  
---  ------             --------------  -----  
 0   Пол                5000 non-null   int64  
 1   Возраст            5000 non-null   float64
 2   Зарплата           5000 non-null   float64
 3   Члены семьи        5000 non-null   int64  
 4   Страховые выплаты  5000 non-null   int64  
dtypes: float64(2), int64(3)
memory usage: 195.4 KB


In [4]:
df.describe()

Пол      Возраст      Зарплата  Члены семьи  Страховые выплаты
count  5000.000000  5000.000000   5000.000000  5000.000000        5000.000000
mean      0.499000    30.952800  39916.360000     1.194200           0.148000
std       0.500049     8.440807   9900.083569     1.091387           0.463183
min       0.000000    18.000000   5300.000000     0.000000           0.000000
25%       0.000000    24.000000  33300.000000     0.000000           0.000000
50%       0.000000    30.000000  40200.000000     1.000000           0.000000
75%       1.000000    37.000000  46600.000000     2.000000           0.000000
max       1.000000    65.000000  79000.000000     6.000000           5.000000

In [5]:
df.duplicated().sum()

153

In [6]:
df.drop_duplicates(inplace=True)
df.duplicated().sum()

0

Все данные на месте и не требуют предобработки.

## Умножение матриц

В этом задании вы можете записывать формулы в *Jupyter Notebook.*

Чтобы записать формулу внутри текста, окружите её символами доллара \\$; если снаружи —  двойными символами \\$\\$. Эти формулы записываются на языке вёрстки *LaTeX.* 

Для примера мы записали формулы линейной регрессии. Можете их скопировать и отредактировать, чтобы решить задачу.

Работать в *LaTeX* необязательно.

Обозначения:

- $X$ — матрица признаков (нулевой столбец состоит из единиц)

- $y$ — вектор целевого признака

- $P$ — матрица, на которую умножаются признаки

- $w$ — вектор весов линейной регрессии (нулевой элемент равен сдвигу)

Предсказания:

$$
a = Xw
$$

Задача обучения:

$$
w = \arg\min_w MSE(Xw, y)
$$

Формула обучения:

$$
w = (X^T X)^{-1} X^T y
$$

**Вопрос:** Признаки умножают на обратимую матрицу. Изменится ли качество линейной регрессии? 

**Ответ:** Не изменится.

**Обоснование:** Для обоснования ответа получим предсказания для домноженных признаков. Пусть наши признаки будут как в исходном датасете размерностью 5000 на 4, а матрица для домножения 4 на 4. Умноженная матрица будет иметь вид X * P. Тогда фомула предсказаний примет следующий вид:

$$
a' = XP'w
$$



$$
a' = X'w' = XPw'
$$

$$
w' = (X'^T X')^{-1} X'^T y = ((XP)^TXP)^{-1})(XP)^Ty
$$

Подставим в формулу предсказаний формулу обучения для матрицы X P:

$$
a' = XP((XP)^TXP)^{-1})(XP)^Ty
$$

Воспользуемся следующими свойствами матриц:

$$
(AB)^{-1} = B^{-1}A^{-1}
$$
$$
(AB)^T = B^TA^T
$$
$$
AA^{-1} = E
$$
$$
AE = EA = A
$$



Раскроем скобки с транспонированием матриц, получится следующее уравнение:

$$
a' = XP(P^TX^TXP)^{-1})P^TX^Ty
$$

Внутри скобок, от множителей которых берется обратная матрица остались следующие матрицы:  $P^T$ размером 4х4, $X^T$ размером 4х5000, $X$ размером 5000x4, $P$ размером 4x4. Одно из дополнительных свойств матриц, о котором следует помнить слудующее $A(BC)=(AB)C$ . Оно позволяет группировать матрицы внутри произведения различным образом. Т.к. обратная матрица может браться только от квадратной матрицы, то для раскрытия скобок данного произведения нам надо сгруппировать его множители так, чтобы они были квадратными матрицами.

Выделим 2 множителя произведения $P^TX^TX$ и P и раскроем скобки:

$$
a' = XPP^{-1}(P^TX^TX)^{-1}P^TX^Ty
$$

Выделим внутри оставшихся скобок множители $P^T$ и $X^TX$ и раскроем скобки:

$$
a' = XPP^{-1}(X^TX)^{-1}(P^T)^{-1}P^TX^Ty
$$

Произведения $PP^{-1}$ и $(P^T)^{-1}P^T$ дают единичные матрицы, которые можно исключить из уровнения. С учетом этого формула принимает вид:

$$
a' = (X^TX)^{-1}X^Ty
$$

Не сложно заметить, что произведение  равно , тогда:

$$
a' = Xw
$$

Таким образом, мы получили аналогичную формулу предсказаний, что и была в начале, значит домножение признаков на обратимую матрицу не влияет на предсказания модели

## Алгоритм преобразования

**Алгоритм**

В качестве алгоритма преобразования примем домножение обучающих признаков X на случайную обратимую матрицу P размером 4х4. Создадим модель линейной регрессии до преобразования и посчитаем метрику R2. Далее умножим исходные признаки на обратимую матрицу и на основе полученных значений вновь посчитаем метрику R2. Метрики R2 по итогу должны быть равны.

**Обоснование**

Как было показано в предыдущем пункте, домножение признаков на какую-либо обратимую матрицу не приводит к изменению предсказаний, поэтому результаты R2 должны оказаться равны.

## Проверка алгоритма

Подготовим необходимые данные: 
- обучающие исходные признаки, 
- целевой признак,
- матрицу для умножения, 
- обучающие признаки после умножения.

После подготовки признаков проведем машинное обучение с помощью алгоритма линейной регрессии для признаков (до умножения и после). Сделаем в каждом случае предсказания и определим метрики r2

In [7]:
features = df.drop('Страховые выплаты', axis=1)
target = df['Страховые выплаты']
display(features.head())
display(target.head())

Пол  Возраст  Зарплата  Члены семьи
0    1     41.0   49600.0            1
1    0     46.0   38000.0            1
2    0     29.0   21000.0            0
3    0     21.0   41700.0            2
4    1     28.0   26100.0            0

0    0
1    1
2    0
3    0
4    0
Name: Страховые выплаты, dtype: int64

Проведем моделирование:

In [8]:
model = LinearRegression()
model.fit(features, target)
predictions = model.predict(features)
r2 = r2_score(target, predictions).round(4)
print(f'Метрика R2 для признаков в исходном состоянии равняется {r2}')

Метрика R2 для признаков в исходном состоянии равняется 0.4302


Теперь необходимо создать обратимую матрицу размером 4 на 4. Достаточно важным является не создать случайно необратимую матрицу, т.к. это нарушит наш алгоритм. 

По одной из теорем обратных матриц: 
- для существования обратной матрицы $A^{-1}$ необходимо и достаточно, чтобы матрица $A$ была невырожденной, то есть, чтобы ее определитель не был равен нулю. 

Заложим проверку данного условия при создании обратимой матрицы.

Зададим исходную точку определителя:

In [9]:
a = 0
b = 0
while a == 0:
    b = np.random.normal(size=(4, 4))
    a = np.linalg.det(b)
b

array([[-0.12614625, -2.24379482, -0.13667589,  0.69664255],
       [-0.44285136, -1.0240766 ,  0.22298126, -0.75667183],
       [ 0.8718741 , -0.78378867,  0.60977603, -0.50248555],
       [ 0.13122245,  0.0840288 ,  1.62989168,  0.49290433]])

Матрица обратима, умножим исходные признаки на нашу матрицу:

In [10]:
features_multi = features @ b
features_multi.head()

0             1             2             3
0  43226.803371 -38920.064719  30255.526541 -24953.117285
1  33110.975738 -29830.992787  23183.376174 -19128.764905
2  18296.513343 -16489.260199  12811.763089 -10574.140036
3  36348.112403 -32705.324906  25435.602846 -20968.551741
4  22743.387942 -20487.802111  15921.261186 -13135.363027

Произведем моделирование на измененных признаках.

In [11]:
model_new = LinearRegression()
model_new.fit(features_multi, target)
predictions = model_new.predict(features_multi)
r2_multi = r2_score(target, predictions).round(4)
print(f'Метрика R2 для признаков в умноженном состоянии равняется {r2_multi}')

Метрика R2 для признаков в умноженном состоянии равняется 0.4302


Метрики R2 обеих моделей одиннаковые, а значит умножение исходных данных на случайную обратимую матрицу может быть применено для их шифрования

## Вывод

В ходе работы был разработан алгоритм, где данные были преобразованы методом умножения признаков исходных данных на обратимую матрицу. Результатом такого преобразования является новая (изменённая) матрица признаков, которая справилась с основными задачами по шифровке данных и при сравнении качества предсказаний сохранила значения метрики r2 неизменной (0.4302).